tools/dissector/sample_dissector.cpp

```diff
@@ -196,37 +193,94 @@ namespace OpenDDS
-      Sample_Manager::instance().add_protocol_field(
-        &fc->hf_, get_ns(), fc->label_, ft, fd
-      );
+    void Sample_Base::push_idl_name(std::string name) {
+      Sample_Base::clear_ns();
+
+      // If name begins with "IDL:", remove it
+      const std::string idl_prefix("IDL:");
+      if (name.substr(0, idl_prefix.size()) == idl_prefix) {
+          name.erase(0, idl_prefix.size());
+      }
+
+      // If name contains ':', remove everything after the last ':'
+      size_t l = name.rfind(":");
+      if (l != std::string::npos) {
+          name.erase(l, name.size() - l);
+      }
+
+      // Push namespace when we find a '/'
+      name.push_back('/'); // For last namespace
+      l = 0;
+      for (size_t index = 0; index != name.size(); index++) {
+          if (name[index] == '/') {
+              Sample_Base::push_ns(name.substr(index - l, l));
+              l = 0;
+          } else {
+              l++;
+          }
+      }
+    }
+
+    void Sample_Base::add_protocol_field() {
+      Field_Context* fc = get_context();
+      if (fc) {
+        Sample_Manager::instance().add_protocol_field(fc->hf_info_);
+      }
     }
 
-    Sample_Field::Sample_Field (IDLTypeID id, const std::string &label)
-      : label_ (label),
-        type_id_ (id),
-        nested_ (0),
+    Field_Context* Sample_Base::create_context(ftenum ft, field_display_e fd) {
+      if (field_contexts_.count(get_ns())) {
+        throw Sample_Dissector_Error(
+          std::string("Could not create context becuase it already exists for: ") +
+          get_ns()
+        );
+      }
+
+      Field_Context* fc = new Field_Context;
+      fc->long_name_ = get_ns();
+      fc->short_name_ = get_label();
+      fc->hf_ = -1;
+      fc->hf_info_ = { &fc->hf_, {
+      // Note: Wireshark calls this second struct member "abbrev" (What we
+      // call long_name_) is the string used for filtering.
+      // In our case (and probably most cases for any WS dissector) "abbrev"
+      // will be longer than the first member "name".
+      // This was confusing to me, so just wanted to make a note of this.
+        fc->short_name_.c_str(), fc->long_name_.c_str(),
+        ft, fd, NULL, 0, NULL, HFILL
+      }};
+      field_contexts_[ns_] = fc;
+
+      return fc;
+    }
+
+    Sample_Field::Sample_Field(IDLTypeID id, const std::string &label)
+      : label_(label),
+        type_id_(id),
+        nested_(0),
         next_(0)
     {
     }
 
-    Sample_Field::Sample_Field (Sample_Dissector *n, const std::string &label)
+    Sample_Field::Sample_Field(Sample_Dissector* n, const std::string &label)
       : label_(label),
-        type_id_ (Undefined),
-        nested_ (n),
+        type_id_(Undefined),
+        nested_(n),
         next_(0)
     {
     }
 
-    Sample_Field::~Sample_Field ()
+    Sample_Field::~Sample_Field()
     {
-      delete nested_;
+      /* This doesn't work since a Dissector can be referenced by multiple
```

Comment:
Does that mean this has a leak?  Or is there some other mechanism already managing it?

Reply:
The Wireshark dissector API doesn't have a cleanup callback for dissectors, at least one that I've seen, so this might have never been called until I started messing with it. The dissector tree lives as long as Wireshark is alive and is not wasted space because it can be used multiple times, so I'm not quite sure if that qualifies as a leak or not...